### PR TITLE
Feature/implement stake feature

### PR DESCRIPTION
**Pull request by gaupoit**
### Summary

- Implement the delegate feature

- Tested cases

**Cases A**: Account doesn't have any delegations

- Delegate the valid amount 
- Delegate the insufficient amount => display error

**Case B:** Account has existing delegations

- Delegate the valid amount to new validator
- Delegate the valid amount to the delegated validator

### Things to do next

- Undelegate

### CODE REVIEW

File: client/src/components/Common/Layout/Stake/Table.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+);
+
+const StakingAccountListComponent = ({ stakingDeployList = [] }) => {
+	if (!stakingDeployList.length) {
+		return <EmptyDelegation />;
+	}
+	const total = stakingDeployList
+		.filter((stake) => !!stake.successAmount)
+		.map((stake) => stake.successAmount)
+		.reduce((prev, next) => prev + next);
```
Comment (djhahe) [inline client/src/components/Common/Layout/Stake/Table.jsx:27]: Please add default value for reduce function. 
It throw error if array is empty
TypeError: Reduce of empty array with no initial value

File: client/src/components/Common/Layout/Stake/Form.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.cd_setting_list_items {
+	.cd_setting_items_form {
```
Comment (djhahe) [inline client/src/components/Common/Layout/Stake/Form.scss:2]: Please define css in public folder for easier  maintenance   later

File: client/src/components/Common/Layout/Stake/Form.jsx [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+	// Selector
+	const { error: deployError, loading: isDeploying } = useSelector(deploySelector);
+
+	const options = validators
+		? validators.map(({ public_key: publicKey, bid }) => ({
+				value: publicKey,
+				label: publicKey,
+				rate: bid.bid.delegation_rate,
+				icon: <i class="bi bi-person"></i>,
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/Form.jsx]: Fixed it: https://github.com/CasperDash/casperdash-client/pull/33/commits/ef4c7d251c480bbba548c77869dd775b98af0675

File: client/src/components/Common/Layout/Stake/Form.scss (hunk 2 above)
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/Form.scss]: Fixed in https://github.com/CasperDash/casperdash-client/pull/33/commits/3662b136840b6db078888abaf496c0da0438471a

File: client/src/components/Common/Layout/Stake/Form.jsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+import Select from 'react-select';
+
+import { useDispatch, useSelector } from 'react-redux';
+
+import ConfirmationModal from './Modal';
+
+import { getSignedStakeDeploy } from '../../../../services/stakeServices';
+import { putDeploy } from '../../../../actions/deployActions';
+import { pushStakeToLocalStorage } from '../../../../actions/stakeActions';
+
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/Form.jsx:13]: I will create another PR to fix it after reviewing the PR https://github.com/CasperDash/casperdash-client/pull/34

**Comment by gaupoit** [comment]
hi @djhahe , I have fixed the table width issue. 

Screen Shot 2021-11-22 at 14 23 34

File: client/src/store/reducers/stakes.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { getLocalStorageValue } from '../../services/localStorage';
+import { DEPLOY } from '../actionTypes';
+
+export default function userReducer(state = { delegations: [] }, action) {
+	switch (action.type) {
+		case DEPLOY.PUSH_STAKE_TO_LOCAL_STORAGE:
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/store/reducers/stakes.js]: Fixed it in [WsNe0Sc2O6K4](https://github.com/CasperDash/casperdash-client/pull/33/commits/0f2bfaf8bc3030b76ecdad40a6f2208ed22bbe6f)

File: client/src/components/Common/Layout/Stake/Table.jsx (hunk 1 above)
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/Table.jsx]: Updated.

File: client/src/services/stakeServices.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+			delegator,
+			validator,
+			amount: CLValueBuilder.u512(amount),
+		},
+		fee,
+	);
+};
+
+export const getSignedStakeDeploy = async ({ fromAddress, validator, fee, amount }) => {
+	const fromAccPk = CLPublicKey.fromHex(fromAddress);
```
Comment (djhahe) [inline client/src/services/stakeServices.js:33]: Please catch this as well, CLPublicKey.fromHex() may throw exception

**Review by djhahe — APPROVED** [review]
looks good, can merge after adding catch for public key convert

**Comment by gaupoit** [comment]
hi @djhahe , plz help me to re-review again. There are some changes:

- [ ]  Implement the separated forms for delegate and undelegate actions 

- [ ] Revamp the delegations table and implement the Undelegate button

- [ ] Handle the bug when user enters the staked amount which has more than 9 decimal places

File: client/src/actions/stakeActions.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	request: {
+		url: '/validators',
+	},
+});
+
+export const updateStakeDeployStatus = (publicKey, path, listHash = []) => {
+	return (dispatch) => {
+		const deployStorageValue = getLocalStorageValue(publicKey, path) || [];
+		const updatedValue = deployStorageValue.map((deploy) => {
+			const hashStatus = listHash.find((hash) => hash.hash.toLowerCase() === deploy.deployHash.toLowerCase());
```
[the remark this answers is not in the document]
Comment (djhahe) [reply client/src/actions/stakeActions.js:27]: Screen Shot 2021-11-28 at 10 56 52

File: client/src/components/Common/Layout/Stake/DelegateForm.jsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+						<ConfirmationModal
+							title="Confirm delegation"
+							show={showModal}
+							validator={stakeDetails.validator}
+							fromAddress={stakeDetails.fromAddress}
+							amount={stakeDetails.amount}
+							fee={stakeDetails.fee}
+							currentPrice={csprPrice}
+							onClose={onCloseModal}
+							onConfirm={onComfirm}
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/DelegateForm.jsx:235]: Thanks, I fixed it https://github.com/CasperDash/casperdash-client/pull/33/commits/37ad72aeb12ffa4c37075d57f34a0a53c912e62b

File: client/src/components/Common/Layout/Stake/UnDelegateForm.jsx [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+										<Button className="cd_send_currency_btn" onClick={handleToggle}>
+											Back
+										</Button>
+										<Button
+											className="cd_send_currency_btn"
+											type="submit"
+											disabled={!values.amount}
+											onClick={handleSubmit}
+										>
+											Send
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/UnDelegateForm.jsx]: I changed in https://github.com/CasperDash/casperdash-client/pull/33/commits/37ad72aeb12ffa4c37075d57f34a0a53c912e62b

**Review by djhahe — APPROVED** [review]
looks good, can merger after fixing some minor issues

File: client/src/components/Common/Layout/Stake/UnDelegateForm.jsx [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+											</div>
+											<Form.Control.Feedback type="invalid">
+												{errors.amount}
+											</Form.Control.Feedback>
+										</div>
+										<div className="cd_send_currency_text_type">
+											{csprPrice && !errors.amount ? (
+												<>
+													<h3 className="cd_send_currency_text">
+														${parseFloat(values.amount * csprPrice).toFixed(2)}
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/UnDelegateForm.jsx]: Fixed in https://github.com/CasperDash/casperdash-client/pull/33/commits/71694fde605ee0077ab6041d3515a119103f0200

File: client/src/components/Common/Layout/Stake/UnDelegateForm.jsx [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+								<span>{validator}</span>
+							</td>
+							<td className="cd_account_table_action">
+								{validator && <CommonAction type="account" value={validator} />}
+							</td>
+						</tr>
+						{info && info.bidInfo && info.bidInfo.bid && (
+							<>
+								<tr>
+									<td>Commision Rate</td>
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/UnDelegateForm.jsx]: Fixed in https://github.com/CasperDash/casperdash-client/pull/33/commits/3ad0fb9fe267caa0e7504a49d98594db60ae7e76

File: client/src/components/Common/Layout/Stake/Table.jsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+					: 'You do not have any delegations yet. Stake CSPR, earn rewards and help Capser become more secure!'}
+			</div>
+		</div>
+	</div>
+);
+
+const TableActions = ({ validator, unDelegateFunc, disableAction }) => {
+	return (
+		<>
+			{disableAction}
```
[the remark this answers is not in the document]
Comment (gaupoit) [reply client/src/components/Common/Layout/Stake/Table.jsx]: Yeah that's rite.

**Pull request merged**